README
(Incomplete) readme for Mac Eve Tools

This project is licenced under the GNU GPL v3.

Building:

Requires Snow Leopard and Xcode 3.2
It can be made to work on Leopard 10.5 with a little effort.

Mac Eve Tools can use Sparkle to manage updates and can be found here:
http://sparkle.andymatuschak.org/

Quasi design document:

The Private/MainController.m file is where the execution starts.
first in init, then awakeFromNib will be called, then appIsActive.
After that the program is up and running and ready for use.

Core:
	Core functionality, the user interface calls these classes to do stuff.
  	Heavy lifting such as calculating skill plans, managing character updates,
	database access and skill tree and stuff is all in here.
Core/Controls
	User interface elements
Core/Character
	The character object and related support classes.
Core/SQLite Database
	SQLite database implementation used for storing skill plans
Core/Skills
	Skill tree and Skill plan related data.
Views:
	The user interface is implemented in here, it will use the Core classes
	to perform its operations.
Windows:
	Pop up windows that display info about a skill, ship, certificate etc.
Private:
	Top level stuff that handles app startup / shutdown and manages the main
	window. 


src/ReleaseNotes.txt
Release 16.17 (22/01/2011)
	Support incursion character portraits.
	Incursion 1.1.0 database.
  
Release 16.16 (18/12/2010)
         Incursion 1.0.1

Release 16.14 (14/07/2010)
	Fixed a bug that caused the wrong skill plan to be deleted

Release 16.13 (13/07/2010)
	Now with 10% more workage!

Release 16.12 (12/07/2010)
	Fixed incorrect training time display for currently training skill.
	Add a display to the skill window that shows what is enabled
		by training this skill.
	Config screen rewrite by Rainor Drawyer.
	Memory leaks plugged.

Release 16.11 (13/06/2010)
	Tyrannis 1.0.1 Database
	Dock icon thanks to leefur eelz.

Release 16.10 (29/05/2010)
	Tyrannis 1.0 Support.

Release 16.9 (11/04/2010)
	The training queue now has a live countdown

Release 16.8 (08/04/2010)
	Fixed learning skills not adjusting training times.
	Attribute calculator.
	Evemon EMP & XML skill plan importing and exporting.
	Character portraits can be updated.

Release 16.7 (01/03/2010)
	German translation provied by eXeler0n
	All skill and certificate information now comes from
		the database.  German and Russian language 
		support from the database. (see preference screen)
	Code cleanup

Release 16.6 (29/01/2010)
	Fixed a bug in displaying the skill details window
	 from the skill plan.
	Provide training time estimate in the ship and cert
	 popup windows.
	Fixed certificate searching.
	 
Release 16.5 (27/01/2010)
	Added certificate support to the skill planner.

Release 16.4 (22/01/2010)
	Ship details window.  provide detailed info on all ships.
		requires new database.
	Database updated.

Release 16.3 (21/12/2009)
	Add option to increase skill level directly from plan (right click)
	Misc bug fixes.

Release 16.2 (23/11/2009) - BETA RELEASE
	Add plus / minus buttons to the skill planner.
	Notes button not implemented.
	BUG: Clicking on the button too fast triggers a double click event.

Release 16.1 (26/10/2009)
	Add option to close skill plans (right click on tab).
	Fix bug caused by not having skill plan columns set.

Release 16 (12/10/2009)
	Rewrite of the character update procedure.
	Lots of internal refactoring.
	Skill plans can be renamed.
	Skill planner column reordering (See preference screen).

Alpha 15.3 (10/9/2009)
	Fix a crash caused by the server status monitor.
	Snow Leopard only release.
	
Alpha 15.1 (31/8/2009)
	Fix the skill queue (bug introduced in 
		release 15)
	Overview now updates when the character  is refreshed
	Known Issues: Toolbar customization 
		fails under leopard (10.5)
Alpha 15 (30/8/2009)
	Character overview page.
	New colour scheme for the training queue.
	Fix a database verification bug under OSX 10.6.
	Server status monitor.
	Window now opens where it was closed last.
	Customisable toolbar.
	CMD-R refresh button shortcut.
	Batch character updates.
		(See prefrence menu to enable)
	Other bug fixes.

Alpha 14.5 (9/8/2009)
	Tweaks to the skill cell to clip overlapping text
	Skill planner properly refreshes on character update

Alpha 14.4 (3/8/2009)
	Fixed a crash on character update during XML parsing.

Alpha 14.3 (2/8/2009)
	New XML update procedure.
	Fixes for the skill queue (Queue would error when the queue was
		completed)

Alpha 14.2 (28/7/2009)
	New Skill queue.
	
Alpha 14.1 (20/7/2009)
	Fixed a crash on startup for new users or accounts.
	Tweaks to the skill cell on the character sheet.

Alpha 14 (18/7/2009)
	New cell for the skill sheet which replicates the in game cell.
	Revamp of the skill planner frontend.

Alpha 13.1 (13/7/2009)
	Fixed a bug that would cause the skill planner to fail to create new plans
		for new accounts.

Alpha 13 (13/7/2009)
	Ship planning. (requires database. MacEveApi->download database in menu bar)
	Shows percent complete of currently training skill in the skill planner.
	Fix a random crash on character update.
	In the skill prerequisite, display skills that are trained but
		not to the requried level as orange
	Fixed a problem caused by closing the skill plan database too soon
	
Alpha 12 (5/7/2009)
	Memory leak fixes.
	Fixed a few stupid bugs.
	Fixed a bug in the skill plan where some skills would 
		not be deleted properly, if a prerequisite was deleted.
	Fixed a bug where the program would freeze up or crash 
		if there were zero skills in the skill plan. 
	
Alpha 10 (27/6/2009)
	Fixed a crash caused by double clicking on empty rows.
	Skill planner now correctly calculates sp/hr adjusted for learning skills.

Alpha 9
	Added automatic update support using the Sparkle Updater (http://sparkle.andymatuschak.org/)
	The skill planner now correctly accounts for learning skills that are added to the queue.
	Fixed a spelling error in the skill details window.
	Minor cosmetic changes to the skill details window.
	BUG: the skill planner does not correctly calculate SP/hr when a learning skill is added to a plan
	
Alpha 8
	Added tooltips to the character skill list and the planner skill list.
	Added support for the training queue.
	Added a skill info window. accesable by right click->info in the skill planner,
		or double click on a skill in the character sheet or skill planner. 
		(Window idea copied from EveHQ)
	Single clicking on a row will expand that row (instead of having to click on the triangle)

Alpha 7 (18/6/2009)
	Fix memory leaks in the skill planner.
	Add start, finish, and percent completed columns in the skill plan.
	Added automatic column resizing to the skill plan.
	Replace the toolbar drop down list with buttons.
	Display clone skill points, set clone SP to red if the clone needs updating.
	Fixed a startup crash related to parsing character XML sheets.

Alpha 6 (17/6/2009)
	Add the ability to search for a skill in the skill plan.
	Fix a bug that would cause 32 bit versions not to display the skill completion countdown properly

Alpha 5
	Remove a skill plan by right click -> remove
	Fix a bug where the skill plan didn't always save 
	Add experimental support for reordering skills in a skill plan

Alpha 4

src/Core/macros.h
/*
 This file is part of Mac Eve Tools.
 
 Mac Eve Tools is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
 Mac Eve Tools is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with Mac Eve Tools.  If not, see <http://www.gnu.org/licenses/>.
 
 Copyright Matt Tyson, 2009.
 */
/*
	Generic preprocessor macros for use throughout the project
 */

/*
 These are the keys for the values stored in the NSMutableDictionary in the Character object
 */

#define TAG_PLUS_BUTTON 0
#define TAG_MINUS_BUTTON 1

#define CHAR_ID @"characterID"
#define CHAR_NAME @"name"
#define CHAR_RACE @"race"
#define CHAR_BLOODLINE @"bloodLine"
#define CHAR_GENDER @"gender"
#define CHAR_CORP_NAME @"corporationName"
#define CHAR_CORP_ID @"corporationID"
#define CHAR_CLONE_NAME @"cloneName"
#define CHAR_CLONE_SP @"cloneSkillPoints"
#define CHAR_BALANCE @"balance"

#define CHAR_TRAINING_END @"trainingEndTime"
#define CHAR_TRAINING_START @"trainingStartTime"
#define CHAR_TRAINING_TYPEID @"trainingTypeID"
#define CHAR_TRAINING_STARTSP @"trainingStartSP"
#define CHAR_TRAINING_ENDSP @"trainingDestinationSP"
#define CHAR_TRAINING_LEVEL @"trainingToLevel"
#define CHAR_TRAINING_SKILLTRAINING @"skillInTraining" 
/*SKILLTRAINING will be @"1" if a skill is training, @"0" if not training*/

#define CHAR_ERROR_CHARSHEET 0
#define CHAR_ERROR_TRAININGSHEET 1
#define CHAR_ERROR_QUEUE 2
//total error sheets
#define CHAR_ERROR_TOTAL 3

/*Attribute macros*/

enum AttributeType {
	attr_intelligence,
	attr_memory,
	attr_charisma,
	attr_perception,
	attr_willpower
};

#define ATTR_INTELLIGENCE 0
#define ATTR_INTELLIGENCE_STR @"intelligence"
#define ATTR_INTELLIGENCE_STR_UPPER NSLocalizedString(@"Intelligence",@"int attribute")
#define ATTR_MEMORY 1
#define ATTR_MEMORY_STR @"memory"
#define ATTR_MEMORY_STR_UPPER NSLocalizedString(@"Memory",@"memory attribute")
#define ATTR_CHARISMA 2
#define ATTR_CHARISMA_STR @"charisma"
#define ATTR_CHARISMA_STR_UPPER NSLocalizedString(@"Charisma","charisma attribute")
#define ATTR_PERCEPTION 3
#define ATTR_PERCEPTION_STR @"perception"
#define ATTR_PERCEPTION_STR_UPPER NSLocalizedString(@"Perception",@"perception attribute")
#define ATTR_WILLPOWER 4
#define ATTR_WILLPOWER_STR @"willpower"
#define ATTR_WILLPOWER_STR_UPPER NSLocalizedString(@"Willpower",@"willpower attribute")

#define ATTR_TOTAL 5

/* Views for the main window */
#define VIEW_CHARSHEET 0
#define VIEW_UPDATE 1
#define VIEW_SKILLPLAN 2


/*table column headers that the SkillTree delegate will respond to*/
#define COL_SKILL_NAME @"SKILL"
#define COL_SKILL_RANK @"RANK"
#define COL_SKILL_POINTS @"POINTS"
#define COL_SKILL_CURLEVEL @"LEVEL"
/*table column identifiers for the skill overview*/
#define COL_POV_NAME @"NAME"
#define COL_POV_SKILLCOUNT @"SKILLCOUNT"
#define COL_POV_TIMELEFT @"TIMELEFT"

/*colums for the skill dependant class*/
#define COL_DEP_NAME @"DEPTYPENAME"
#define COL_DEP_SKILLLEVEL @"DEPSKILLLEVEL"

/*colums the skill plan view will respond to*/
#define COL_PLAN_CALSTART @"CALSTART"
#define COL_PLAN_CALFINISH @"CALFINISH"
#define COL_PLAN_PERCENT @"%"
#define COL_PLAN_SKILLNAME @"SKILLNAME"
#define COL_PLAN_SPHR @"SPHR"
#define COL_PLAN_TRAINING_TIME @"TRAININGTIME"
#define COL_PLAN_TRAINING_TTD @"TIMETODATE"
#define COL_PLAN_BUTTONS @"BUTTONS"
#define COL_PLAN_ATTRIBUTES @"ATTRIBUTES"
#define COL_PLAN_POINTS COL_SKILL_POINTS


/*XML Documents we are interested in downloading*/
#define XMLAPI_CHAR_SHEET @"/char/CharacterSheet.xml.aspx"
#define XMLAPI_CHAR_LIST @"/account/Characters.xml.aspx"
#define XMLAPI_CHAR_TRAINING @"/char/SkillInTraining.xml.aspx"
#define XMLAPI_CHAR_QUEUE @"/char/SkillQueue.xml.aspx"
#define XMLAPI_SKILL_TREE @"/eve/SkillTree.xml.aspx"
#define XMLAPI_WALLET_REF @"/eve/RefTypes.xml.aspx"
#define XMLAPI_SERVER_STATUS @"/server/ServerStatus.xml.aspx"
#define XMLAPI_CERT_TREE @"/eve/CertificateTree.xml.aspx"

/*the portrait will be saved to this file in the characters directory*/
#define PORTRAIT @"portrait.jpg"

#define MIN_DAY 1440
#define MIN_HOUR 60
#define MIN_MINUTE 1

#define SEC_DAY 86400
#define SEC_HOUR 3600
#define SEC_MINUTE 60

#define MTSkillPairPBoardType @"MTSkillPairPBoardType"
#define MTSkillArrayPBoardType @"MTSkillArrayPBoardType"

#define MTSkillIndexPBoardType @"MTSkillIndexPBoardType"

#define DRAG_TYPEID @"typeid"
#define DRAG_SKILLLEVEL @"level"

#define DRAG_SKILLINDEX @"skillindex"

/*Skill details pop up window*/
#define SD_LEVEL @"TT_LEVEL"
#define SD_TIME @"TT_TIME"
#define SD_TOTAL @"TT_RUN_TOTAL"
#define SD_FROM_NOW @"TT_FROM_NOW"
#define SD_SP_LEVEL @"TT_SP_LEVEL"
#define SD_SP_DIFF @"TT_SP_DIFF"

#define SD_PREREQUISITE @"SD_PRE"

#define GROUP_LEARNING 267

#define SKILL_LEARNING 3374
#define SKILL_MEMORY 3377


// Max counts

#define COUNT_CATEGORY 32
#define COUNT_GROUPS 826;


/*bonuses*/

/*These magic numbers are the attribute IDs as they appear in the
 CCP database export*/
#define BONUS_LEARNING 455
#define BONUS_INTELLIGENCE 176
#define BONUS_CHARISMA 175
#define BONUS_PERCEPTION 178
#define BONUS_WILLPOWER 179
#define BONUS_MEMORY 177

/*Notifications*/

#define NOTE_DATABASE_DOWNLOAD_COMPLETE @"databaseDownloadCompleted"

/*
 the character sheet has finished updating. skill plans may or may not have
 been purged
 */
#define CHARACTER_SHEET_UPDATE_NOTIFICATION @"MTCharacterSheetDidUpdate"

#define SERVER_STATUS_NOTIFICATION @"MTServerStatusNotification"

#define CHARACTER_SKILL_PLAN_PURGED @"MTSkillPlanPurged"

/*CCP Database*/

#define DB_CATEGORY_SHIP 6
#define DB_CATEGORY_MODULE 7
#define DB_CATEGORY_CHARGE 8
#define DB_CATEGORY_SKILL 16

/* Translation Columns */

#define TRN_TYPE_DESCRIPTION 33
#define TRN_TYPE_NAME 8
#define TRN_GROUP_NAME 7

#define TRN_CRTCRT_DESCRIPTION 107
#define TRN_CRTCLS_NAME 106
#define TRN_CRTCAT_NAME 105


/*NSUserDefault keys*/
#define SKILL_PLAN_CONFIG @"skill_plan_config"
#define UD_DATABASE_LANG @"ud_db_lang"

#define UD_API_URL @"apiUrl"
#define UD_ROOT_PATH @"rootPath"
#define UD_ITEM_DB_PATH @"itemDbPath"
#define UD_SUBMIT_STATS @"submitAnonymousStatistics"
#define UD_CHECK_FOR_UPDATES @"checkForAutomaticUpdates"
#define UD_PICTURE_URL @"picUrl"
#define UD_UPDATE_FEED_URL @"updateFeedUrl"
#define UD_DB_UPDATE_URL @"dbUpdateUrl"
#define UD_DB_SQL_URL @"dbSQLUrl"
#define UD_IMAGE_URL @"imageUrl"
#define UD_DATABASE_MIN_VERSION @"databaseMinimumVersion"

#define UD_ACCOUNTS @"accounts"

enum CCPRace
{
	NullRace = 0,
	Caldari = 1,
	Minmatar = 2,
	Amarr = 4,
	Gallente = 8,
	Jove = 16,
	Pirate = 32
};
typedef enum CCPRace CCPRace;

enum CCPMetaGroup
{
	NullType = 0,
	TechI = 1,
	TechII = 2,
	Storyline = 3,
	Faction = 4,
	Officer = 5,
	Deadspace = 6,
	Frigates = 7,
	EliteFrigates = 8,
	Commander = 9,
	Destroyer = 10,
	Cruiser = 11,
	EliteCruiser = 12,
	CommanderCruiser = 13,
	TechIII = 14
};
typedef enum CCPMetaGroup CCPMetaGroup;


enum ServerStatus
{
	ServerUp,
	ServerDown,
	ServerUnknown,
	ServerStarting
};
typedef enum ServerStatus ServerStatus;

enum StatusImageState
{
	StatusGreen,
	StatusRed,
	StatusGray,
	StatusYellow,
	StatusHidden
};
typedef enum StatusImageState StatusImageState;

enum ImageSize
{
	_16 = 16,
	_32 = 32,
	_64 = 64,
	_128 = 128,
	_256 = 256
};
typedef enum ImageSize ImageSize;

enum AttributeTypeGroups
{
	Drones = 1,
	Structure = 2,
	Armour = 3, 
	Shield = 4,
	Cap = 5,
	Targeting = 6,
	Propulsion = 7,
	Other = 8,
	Fitting = 9
};
typedef enum AttributeTypeGroups AttributeTypeGroups;




/*Don't fuck with these values unless you're special*/
enum DatabaseLanguage
{
	l_EN = 0,
	l_DE = 1,
	l_RU = 2
};
typedef enum DatabaseLanguage DatabaseLanguage;
/*
 macros for sqlite. we use NSIntegers which are defined as longs and therefore change size depending on what platform we are building for
 may need to look at changing this if we ever store anything that could go > MAX_INT. but nothing does at the moment (14/6/09)
 */
#ifdef __LP64__
	#define sqlite3_column_nsint(x,y) sqlite3_column_int64((x),(y))
	#define sqlite3_bind_nsint(x,y,z) sqlite3_bind_int64((x),(y),(z))
	#define sqlite3_column_cgfloat(x,y) sqlite3_column_double((x),(y))
#else
	#define sqlite3_column_nsint(x,y) sqlite3_column_int((x),(y))
	#define sqlite3_bind_nsint(x,y,z) sqlite3_bind_int((x),(y),(z))
	#define sqlite3_column_cgfloat(x,y) (CGFloat)sqlite3_column_double((x),(y))
#endif

/*
 macro hackery for 32 bit platforms
 
 CGFloat is a float on 32 bit platforms, and double on 64 bit platforms
 
 floating point constants follow the same rule, these macros wrap to the
 proper function call for the type we are currently building for, to avoid
 compiler complaints.
 
 the test is copied from CGBase.h where it sets CGFloat to float or double
 */

#if defined(__LP64__) && __LP64__
	#define xround(x) round((x))
	#define xfloor(x) floor((x))
	#define xceil(x) ceil((x))
	#define xlround(x) lround((x))
	#define xfabs(x) fabs((x))
#else
	#define xround(x) roundf((x))
	#define xfloor(x) floorf((x))
	#define xceil(x) ceilf((x))
	#define xlround(x) lroundf((x))
	#define xfabs(x) fabsf((x))
#endif

///http://api.eve-online.com/char/CharacterSheet.xml.aspx?userID=<userid>&apiKey=<apikey>&characterID=<characterID>
///http://api.eve-online.com/char/CharacterSheet.xml.aspx?userID=<userid>&apiKey=<apikey>&characterID=<characterID>
